File: platform/windows/MihomoMeter.Windows.Tests/SharedCoreProxyTypeLazyRouteTests.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using MihomoMeter.Windows.Core.Application;
using MihomoMeter.Windows.Core.Domain;

namespace MihomoMeter.Windows.Tests;

[TestClass]
public sealed class SharedCoreProxyTypeLazyRouteTests
{
    private static readonly ProxyClassification Proxy = new(TrafficCategory.Proxy);
    private static readonly ProxyClassification Direct = new(TrafficCategory.Direct);
    private static readonly ProxyClassification Reject = new(TrafficCategory.Reject);
    private static readonly ProxyClassification Unknown = new(
        TrafficCategory.Unknown,
        UnknownTrafficReason.AmbiguousProxyType);

    [TestMethod]
    public void LazyRouterDoesNotEvaluateFallbackOnSharedSuccess()
    {
        (SharedProxyTypeClassification Shared, ProxyClassification Expected)[] cases =
        [
            (SharedProxyTypeClassification.Proxy, Proxy),
            (SharedProxyTypeClassification.Direct, Direct),
            (SharedProxyTypeClassification.Reject, Reject),
        ];

        foreach (var testCase in cases)
        {
            var fallbackCount = 0;
            var classifyCount = 0;
            var result = SharedCoreProxyTypeRouter.RouteLazy(
                "Synthetic",
                () =>
                {
                    fallbackCount += 1;
                    return Unknown;
                },
                _ =>
                {
                    classifyCount += 1;
                    return testCase.Shared;
                });

            Assert.AreEqual(testCase.Expected, result.Classification);
            Assert.AreEqual(SharedCoreProxyTypeRouteSource.SharedPrimary, result.Source);
            Assert.AreEqual(SharedCoreProxyTypeRouteStatus.Succeeded, result.Status);
            Assert.AreEqual(1, classifyCount);
            Assert.AreEqual(0, fallbackCount);
        }
    }

    [TestMethod]
    public void LazyRouterEvaluatesFallbackExactlyOnceForUnrecognizedAndSharedFailures()
    {
        AssertLazyFallback(
            SharedCoreProxyTypeRouteStatus.Unrecognized,
            _ => SharedProxyTypeClassification.Unrecognized);

        (SharedProxyTypeAdapterFailure Failure, SharedCoreProxyTypeRouteStatus Status)[] failures =
        [
            (SharedProxyTypeAdapterFailure.UnsupportedAbiVersion,
                SharedCoreProxyTypeRouteStatus.AbiMismatch),
            (SharedProxyTypeAdapterFailure.NativeCallFailed,
                SharedCoreProxyTypeRouteStatus.NativeCallFailed),
            (SharedProxyTypeAdapterFailure.UnsupportedProxyTypeInput,
                SharedCoreProxyTypeRouteStatus.UnsupportedInput),
            (SharedProxyTypeAdapterFailure.ProxyTypeInputTooLong,
                SharedCoreProxyTypeRouteStatus.InputTooLong),
            (SharedProxyTypeAdapterFailure.UnsupportedProxyTypeCategory,
                SharedCoreProxyTypeRouteStatus.UnexpectedResult),
        ];
        foreach (var failure in failures)
        {
            AssertLazyFallback(
                failure.Status,
                _ => throw new SharedProxyTypeAdapterException(
                    failure.Failure,
                    "synthetic"));
        }

        AssertLazyFallback(
            SharedCoreProxyTypeRouteStatus.NativeCallFailed,
            _ => throw new DllNotFoundException("synthetic"));
        AssertLazyFallback(
            SharedCoreProxyTypeRouteStatus.UnknownFailure,
            _ => throw new ArgumentException("synthetic"));
    }

    [TestMethod]
    public void LazyRouteIgnoresReporterFailureWithoutEvaluatingFallback()
    {
        var observations = new List<SharedCoreProxyTypeRouteObservation>();
        var fallbackCount = 0;
        SharedCoreProxyTypeRoute.ConfigureReporter(observation =>
        {
            observations.Add(observation);
            throw new InvalidOperationException("synthetic");
        });
        try
        {
            for (var index = 0; index < 2; index += 1)
            {
                Assert.AreEqual(
                    Proxy,
                    SharedCoreProxyTypeRoute.ResolveLazy(
                        "Vmess",
                        () =>
                        {
                            fallbackCount += 1;
                            return Unknown;
                        },
                        _ => SharedProxyTypeClassification.Proxy));
            }

            Assert.AreEqual(0, fallbackCount);
            CollectionAssert.AreEqual(
                new[]
                {
                    new SharedCoreProxyTypeRouteObservation(
                        SharedCoreProxyTypeRouteSource.SharedPrimary,
                        SharedCoreProxyTypeRouteStatus.Succeeded),
                },
                observations);
        }
        finally
        {
            SharedCoreProxyTypeRoute.ConfigureReporter(null);
        }
    }

    private static void AssertLazyFallback(
        SharedCoreProxyTypeRouteStatus expectedStatus,
        Func<string, SharedProxyTypeClassification> classifyProxyType)
    {
        var fallbackCount = 0;
        var classifyCount = 0;
        var result = SharedCoreProxyTypeRouter.RouteLazy(
            "Synthetic",
            () =>
            {
                fallbackCount += 1;
                return Unknown;
            },
            rawType =>
            {
                classifyCount += 1;
                return classifyProxyType(rawType);
            });

        Assert.AreEqual(Unknown, result.Classification);
        Assert.AreEqual(SharedCoreProxyTypeRouteSource.NativeFallback, result.Source);
        Assert.AreEqual(expectedStatus, result.Status);
        Assert.AreEqual(1, classifyCount);
        Assert.AreEqual(1, fallbackCount);
    }
}

File: Tests/SharedCoreProxyTypeLazyRouteTests.swift
import XCTest
import os

@testable import MihomoMeter

final class SharedCoreProxyTypeLazyRouteTests: XCTestCase {
  private let proxy = ProxyClassification(category: .proxy, unknownReason: nil)
  private let direct = ProxyClassification(category: .direct, unknownReason: nil)
  private let reject = ProxyClassification(category: .reject, unknownReason: nil)
  private let unknown = ProxyClassification(
    category: .unknown,
    unknownReason: .ambiguousProxyType
  )

  func testLazyRouterDoesNotEvaluateFallbackOnSharedSuccess() {
    let cases: [(SharedProxyTypeClassification, ProxyClassification)] = [
      (.proxy, proxy),
      (.direct, direct),
      (.reject, reject),
    ]

    for (sharedClassification, expectedClassification) in cases {
      var fallbackCount = 0
      var classifyCount = 0
      let result = SharedCoreProxyTypeRouter.routeLazy(
        rawType: "Synthetic",
        nativeFallback: {
          fallbackCount += 1
          return self.unknown
        },
        classifyProxyType: { _ in
          classifyCount += 1
          return sharedClassification
        }
      )

      XCTAssertEqual(result.classification, expectedClassification)
      XCTAssertEqual(result.source, .sharedPrimary)
      XCTAssertEqual(result.status, .succeeded)
      XCTAssertEqual(classifyCount, 1)
      XCTAssertEqual(fallbackCount, 0)
    }
  }

  func testLazyRouterEvaluatesFallbackExactlyOnceForUnrecognizedAndSharedFailures() {
    assertLazyFallback(expectedStatus: .unrecognized) { _ in .unrecognized }

    let failures: [(SharedProxyTypeAdapterError, SharedCoreProxyTypeRouteStatus)] = [
      (.unsupportedABIVersion(2), .abiMismatch),
      (.nativeCallFailed(-1), .nativeCallFailed),
      (.unsupportedProxyTypeInput, .unsupportedInput),
      (.proxyTypeInputTooLong, .inputTooLong),
      (.unsupportedProxyTypeCategory(99), .unexpectedResult),
    ]
    for (error, status) in failures {
      assertLazyFallback(expectedStatus: status) { _ in throw error }
    }
    assertLazyFallback(expectedStatus: .unknownFailure) { _ in
      throw SyntheticLazyRouteError()
    }
  }

  func testLazyRouteIgnoresReporterFailureWithoutEvaluatingFallback() {
    let observations = OSAllocatedUnfairLock(
      initialState: [SharedCoreProxyTypeRouteObservation]()
    )
    let fallbackCount = OSAllocatedUnfairLock(initialState: 0)
    let fallbackClassification = unknown
    let expectedClassification = proxy
    SharedCoreProxyTypeRoute.configure { observation in
      observations.withLock { $0.append(observation) }
      throw SyntheticLazyRouteError()
    }
    defer {
      SharedCoreProxyTypeRoute.configure(reporter: nil)
    }

    for _ in 0..<2 {
      let result = SharedCoreProxyTypeRoute.resolveLazy(
        rawType: "Vmess",
        nativeFallback: {
          fallbackCount.withLock { $0 += 1 }
          return fallbackClassification
        },
        classifyProxyType: { _ in .proxy }
      )
      XCTAssertEqual(result, expectedClassification)
    }

    XCTAssertEqual(fallbackCount.withLock { $0 }, 0)
    XCTAssertEqual(
      observations.withLock { $0 },
      [
        SharedCoreProxyTypeRouteObservation(
          source: .sharedPrimary,
          status: .succeeded
        )
      ]
    )
  }

  private func assertLazyFallback(
    expectedStatus: SharedCoreProxyTypeRouteStatus,
    classifyProxyType: (String) throws -> SharedProxyTypeClassification,
    file: StaticString = #filePath,
    line: UInt = #line
  ) {
    var fallbackCount = 0
    var classifyCount = 0
    let result = SharedCoreProxyTypeRouter.routeLazy(
      rawType: "Synthetic",
      nativeFallback: {
        fallbackCount += 1
        return self.unknown
      },
      classifyProxyType: { rawType in
        classifyCount += 1
        return try classifyProxyType(rawType)
      }
    )

    XCTAssertEqual(result.classification, unknown, file: file, line: line)
    XCTAssertEqual(result.source, .nativeFallback, file: file, line: line)
    XCTAssertEqual(result.status, expectedStatus, file: file, line: line)
    XCTAssertEqual(classifyCount, 1, file: file, line: line)
    XCTAssertEqual(fallbackCount, 1, file: file, line: line)
  }
}

private struct SyntheticLazyRouteError: Error {}
